Add nightly test CI monitor workflow

alisonshao opened the pull request:
The nightly monitor:
  - Tracks all performance metrics (throughput, latency, accuracy) over time
  - Compares with historical data from the repository
  - Detects regressions >10% automatically
  - Reports everything in GitHub Actions summary with a clean table
  - Saves all data as artifacts for analysis

Update CI monitors to support new nightly workflow structure
Support tracking of new hardware-specific nightly workflows (NVIDIA, AMD, Intel) alongside existing workflows.

Changes:
- Add new NVIDIA job names from nightly-test-nvidia.yml
- Add AMD job names from nightly-test-amd.yml
- Update nightly_monitor.py to fetch from multiple workflow files instead of just nightly-test.yml
- Maintain backward compatibility with old job names

This enables the CI monitor to track failures from all new workflows, making it safe to eventually disable the old nightly-test.yml and nightly-test-b200.yml workflows.

alisonshao committed:
Add nightly test CI monitor workflow

Add dedicated monitoring for nightly test runs to track performance
and accuracy regressions over time:

- New nightly_monitor.py script that analyzes nightly test workflow runs
  - Tracks job success/failure rates
  - Calculates average duration per job
  - Detects high failure rates (>30%)
  - Identifies consecutive failures
  - Generates daily trend reports

- New nightly-monitor.yml workflow that runs daily at 8 AM UTC
  - Analyzes last 7 days of nightly test runs
  - Uploads detailed statistics as artifacts
  - Reports regressions via GitHub Actions output
  - Can be triggered manually with custom date range

- Updated ci_analyzer.py to include nightly-test-8-gpu-b200 job
  in tracking list for general CI monitoring

alisonshao committed:
Fix lint errors in nightly_monitor.py

alisonshao committed:
Fix additional lint errors (line length)

alisonshao committed:
Fix remaining lint formatting issues

alisonshao committed:
Make nightly_monitor.py executable

alisonshao committed:
Add performance metric tracking to nightly monitor (Step 1)

- Add regex patterns for parsing performance metrics from logs
- Add get_job_logs() method to fetch job logs from GitHub API
- Add parse_metrics_from_logs() to extract metrics using regex
- Track performance metrics (throughput, latency, ttft, accuracy) with timestamps
- Display average metrics in report output
- Only fetch metrics from successful perf/eval jobs

This is Step 1 of incremental enhancement to add day-to-day performance
comparison and anomaly detection.

alisonshao committed:
Add historical data fetching from sglang-bot/sglang-ci-data (Step 2)

- Add data_repo and data_branch configuration for sglang-bot/sglang-ci-data
- Add get_historical_data_paths() to list available historical data files
- Add fetch_historical_data() to fetch and decode specific data files from repo
- Add get_recent_historical_metrics() to retrieve metrics for a job over time
- Fetch up to 14 recent historical files to compare with current metrics
- Handle base64 decoding and JSON parsing with error handling

This enables day-to-day metric comparison in the next step.

alisonshao committed:
Apply black formatting to nightly_monitor.py

alisonshao committed:
Add day-to-day metric comparison (Step 3)

- Compare current metrics with 7-day historical average
- Calculate percentage changes and classify as stable/minor/significant
- Display metrics with change indicators and percentages in report

alisonshao committed:
Add >10% anomaly detection (Step 4)

- Detect performance regressions exceeding 10% change
- Flag throughput decreases >10% and latency increases >10%
- Add performance_regression type to regression reports
- Display regression details with current vs 7-day average

alisonshao committed:
Add GitHub summary and artifact generation (Steps 5&6)

- Generate GitHub Actions step summary with regression alerts
- Display performance metrics table with percentage changes
- Highlight regressions detected section
- JSON artifact already created by workflow for all tracked data

alisonshao committed:
Add temporary test workflow for nightly monitor

alisonshao committed:
Trigger test workflow

alisonshao committed:
Update CI monitors to support new nightly workflow structure

Support tracking of new hardware-specific nightly workflows (NVIDIA, AMD, Intel) alongside existing workflows.

Changes:
- Add new NVIDIA job names from nightly-test-nvidia.yml
- Add AMD job names from nightly-test-amd.yml
- Update nightly_monitor.py to fetch from multiple workflow files instead of just nightly-test.yml
- Maintain backward compatibility with old job names

This enables the CI monitor to track failures from all new workflows, making it safe to eventually disable the old nightly-test.yml and nightly-test-b200.yml workflows.

alisonshao committed:
Merge branch 'main' into add-nightly-ci-monitor

alisonshao committed:
Use built-in GITHUB_TOKEN for test workflow

The GH_PAT_FOR_NIGHTLY_CI_DATA secret may not be available in PR/branch contexts. Using the built-in GITHUB_TOKEN instead for testing purposes.

alisonshao committed:
Use GH_PAT_FOR_NIGHTLY_CI_DATA token in test-nightly-monitor

alisonshao committed:
Consolidate nightly monitoring into ci_analyzer.py

- Integrate nightly monitoring functionality into ci_analyzer.py with --mode flag
- Update nightly job names for Nvidia, AMD, and Intel workflows
- Remove standalone nightly_monitor.py file
- Update workflow files to use ci_analyzer.py --mode nightly

alisonshao committed:
Apply black formatting to ci_analyzer.py

alisonshao committed:
Remove separate nightly monitor workflow files

Nightly monitoring is now integrated into ci_analyzer.py with --mode flag

alisonshao committed:
Add nightly monitoring job to ci-monitor workflow

alisonshao committed:
Add performance metrics tracking to existing CI monitor

- Track throughput, latency, accuracy metrics from nightly test jobs
- Parse metrics from successful nightly job logs
- Detect performance trends (>10% changes)
- Display metrics table in GitHub Actions summary with trend indicators
- Save all metrics data in JSON artifacts for analysis

alisonshao committed:
Apply black formatting

alisonshao committed:
Fix 403 error by not filtering by branch by default

The GitHub API requires higher permissions when filtering by branch.
Change default from 'main' to None to avoid 403 Forbidden errors.

alisonshao committed:
Use default GITHUB_TOKEN instead of custom PAT

The custom PAT may not be available on non-main branches.
The default GITHUB_TOKEN has Actions: read permission.

alisonshao committed:
Revert to using GH_PAT_FOR_NIGHTLY_CI_DATA for all steps

Match the main branch configuration. The PAT works on main branch,
issue is likely related to secret access on non-main branches.

alisonshao committed:
Both regular CI analysis and nightly analysis now run together

alisonshao committed:
Remove deprecated nightly-test.yml workflow and old job names from CI analyzer

Stop tracking old nightly jobs that have been replaced by hardware-specific workflows:
- Removed nightly-test.yml from monitored workflows
- Removed old job names (nightly-test-eval-text-models, nightly-test-perf-text-models, etc.)
- Now only tracking new NVIDIA/AMD/Intel specific job names

Kangyan-Zhou reviewed (commented):
Fetching nightly test runs from the last 2 days...
  Fetching from nightly-test.yml...
    Fetched 1000 runs from nightly-test.yml
  Fetching from nightly-test-nvidia.yml...
    Fetched 1000 runs from nightly-test-nvidia.yml
  Fetching from nightly-test-amd.yml...
    Fetched 1000 runs from nightly-test-amd.yml
  Fetching from nightly-test-intel.yml...
    Fetched 1000 runs from nightly-test-intel.yml
Total nightly runs fetched: 4000

Do we need to fetch 4k runs from nightly tests from the last 2 days? This seems a bit off

alisonshao committed:
Limit nightly runs fetching to avoid excessive API calls

Nightly tests run once per day, so fetching 1000 runs per workflow was excessive.
Now limited to fetch at most (days * 5) runs per workflow with smaller page size.
For 2 days: ~10 runs per workflow instead of 1000

alisonshao committed:
Add GitHub Actions summary for nightly test analysis

Now nightly analysis will appear in the GitHub Actions summary showing:
- Overall statistics with daily trends table
- Job statistics with performance metrics
- Recent failures for each job

The daily trends table shows the date and success rate for easy tracking

alisonshao committed:
Fix lint errors - break long lines in nightly summary generation

alisonshao committed:
Apply black formatting

BBuf commented on scripts/ci_monitor/ci_analyzer.py:
```diff
@@ -512,46 +635,567 @@ def generate_github_summary(self, stats: Dict):
+                sys.exit(1)
+            else:
+                print("\n✓ No significant regressions detected")
+                sys.exit(0)
```
`sys.exit(0)` will block the ci-monitor?  https://github.com/sgl-project/sglang/actions/runs/19555793437

alisonshao replied:
checking the issue
